File: examples/compose/duis01_large_component_bad.kt
package examples.compose

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

/**
 * DUIS-01 – LARGE COMPONENT
 *
 * One composable (LargeComponentDashboard) is responsible for:
 *  - layout
 *  - navigation (tabs)
 *  - data loading
 *  - filtering
 *  - simple statistics
 *  - all UI rendering
 *
 * This is a "God component" / large component smell.
 */
@Composable
fun LargeComponentDashboard() {
    var tab by remember { mutableStateOf("home") }
    var name by remember { mutableStateOf("") }
    var items by remember { mutableStateOf(listOf<Int>()) }
    var filter by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(Unit) {
        loading = true
        error = null
        try {
            kotlinx.coroutines.delay(500) // fake API call
            items = (0 until 50).toList()
        } catch (e: Exception) {
            error = "Failed to load items"
        } finally {
            loading = false
        }
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text(text = "Dashboard – $name") }
            )
        }
    ) { padding ->
        Row(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
        ) {
            // Sidebar/navigation + user info
            Column(
                modifier = Modifier
                    .width(220.dp)
                    .padding(16.dp)
            ) {
                Text(text = "User: ${if (name.isBlank()) "(anonymous)" else name}")
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text("Name") },
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(Modifier.height(16.dp))
                Text(text = "Tabs:")
                Row {
                    TextButton(onClick = { tab = "home" }) { Text("Home") }
                    TextButton(onClick = { tab = "stats" }) { Text("Stats") }
                    TextButton(onClick = { tab = "settings" }) { Text("Settings") }
                }
            }

            // Main content – all in this one huge composable
            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(16.dp)
            ) {
                when (tab) {
                    "home" -> {
                        OutlinedTextField(
                            value = filter,
                            onValueChange = { filter = it },
                            label = { Text("Filter") },
                            modifier = Modifier.fillMaxWidth()
                        )
                        Spacer(Modifier.height(8.dp))

                        when {
                            loading -> {
                                CircularProgressIndicator()
                            }

                            error != null -> {
                                Text(text = error!!, color = MaterialTheme.colorScheme.error)
                            }

                            else -> {
                                val filtered = items.filter {
                                    it.toString().contains(filter)
                                }
                                LazyColumn {
                                    items(filtered) { i ->
                                        Text(text = "Item $i", modifier = Modifier.padding(4.dp))
                                    }
                                }
                            }
                        }
                    }

                    "stats" -> {
                        val total = items.sum()
                        val count = items.size
                        Text(text = "Number of items: $count")
                        Text(text = "Sum of items: $total")
                    }

                    "settings" -> {
                        Text("Settings (many fields could go here...)")
                    }
                }
            }
        }
    }
}

File: examples/swiftui/duis01_large_component_bad.swift
import SwiftUI

/// DUIS-01 – LARGE COMPONENT
///
/// One view handles:
///  - layout
///  - navigation (tabs)
///  - data fetching
///  - filtering
///  - simple statistics
///  - all UI rendering
///
/// This is a "God View" / Large Component smell.

struct LargeComponentDashboardView: View {
    @State private var tab: String = "home"
    @State private var name: String = ""
    @State private var items: [Int] = []
    @State private var filter: String = ""
    @State private var loading: Bool = false
    @State private var error: String? = nil

    var body: some View {
        HStack(spacing: 0) {
            // Sidebar
            VStack(alignment: .leading, spacing: 16) {
                Text("Dashboard")
                    .font(.title)
                Text("User: \(name.isEmpty ? "(anonymous)" : name)")
                TextField("Name", text: $name)
                    .textFieldStyle(.roundedBorder)

                Text("Tabs")
                    .font(.headline)
                HStack {
                    Button("Home") { tab = "home" }
                    Button("Stats") { tab = "stats" }
                    Button("Settings") { tab = "settings" }
                }

                Spacer()
            }
            .frame(width: 220)
            .padding()
            .background(Color(white: 0.95))

            Divider()

            // Main content – everything in one big view
            VStack(alignment: .leading, spacing: 16) {
                if tab == "home" {
                    Text("Items")
                        .font(.title2)
                    TextField("Filter", text: $filter)
                        .textFieldStyle(.roundedBorder)
                        .frame(maxWidth: 300)

                    if loading {
                        ProgressView("Loading...")
                    } else if let error = error {
                        Text(error)
                            .foregroundColor(.red)
                    } else {
                        List(filteredItems, id: \.self) { item in
                            Text("Item \(item)")
                        }
                    }
                } else if tab == "stats" {
                    Text("Statistics")
                        .font(.title2)
                    Text("Item count: \(items.count)")
                    Text("Sum: \(items.reduce(0, +))")
                } else if tab == "settings" {
                    Text("Settings")
                        .font(.title2)
                    Text("Imagine a long list of settings here...")
                }

                Spacer()
            }
            .padding()
        }
        .onAppear {
            loadItems()
        }
    }

    private var filteredItems: [Int] {
        if filter.isEmpty { return items }
        return items.filter { "\($0)".contains(filter) }
    }

    // Data fetching + business-ish logic all here
    private func loadItems() {
        loading = true
        error = nil
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            let data = Array(0..<50)
            DispatchQueue.main.async {
                self.items = data
                self.loading = false
            }
        }
    }
}
